File: 189. Rotate Array/Rotate Array.rb
# @param {Integer[]} nums
# @param {Integer} k
# @return {Void} Do not return anything, modify nums in-place instead.
def rotate(nums, k)
	k =  k % nums.size
	return if k === 0
    reverse(nums, nums.size - k, nums.size)
    reverse(nums, 0, nums.size - k)
    reverse(nums, 0, nums.size)
end

def reverse(arr, start_index, end_index)
	((end_index - start_index) / 2).times do |i|
		arr[start_index + i], arr[end_index - i - 1] = arr[end_index - i - 1], arr[start_index + i]
	end
end

File: 189. Rotate Array/Rotate Array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
    k = k % nums.length;
    if(k===0) return;

    reverse(nums, 0, nums.length - k);
    reverse(nums, nums.length - k, nums.length);
    reverse(nums, 0, nums.length);
};

function reverse(arr, start_index, end_index){
	for(var i = 0; i < parseInt((end_index - start_index) / 2); i++) {
		arr[start_index + i] ^= arr[end_index - 1 - i];
		arr[end_index - 1 - i] ^= arr[start_index + i];
		arr[start_index + i] ^= arr[end_index - 1 - i];
	}
}
